File: api/src/main/java/gg/mineral/api/entity/meta/EntityMetadataType.java
package gg.mineral.api.entity.meta;

import java.util.Vector;

import gg.mineral.api.inventory.item.ItemStack;
import gg.mineral.api.math.EulerAngle;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

/**
 * The types of values that entity metadata can contain.
 */
@RequiredArgsConstructor
@Getter
public enum EntityMetadataType {
    BYTE(Byte.class),
    SHORT(Short.class),
    INT(Integer.class),
    FLOAT(Float.class),
    STRING(String.class),
    ITEM(ItemStack.class),
    VECTOR(Vector.class),
    EULER_ANGLE(EulerAngle.class);

    private final Class<?> dataType;

    public static EntityMetadataType byId(int id) {
        return values()[id];
    }

    public byte getId() {
        return (byte) ordinal();
    }
}

File: api/src/main/java/gg/mineral/api/entity/meta/EntityMetadataType.kt
package gg.mineral.api.entity.meta

import gg.mineral.api.inventory.item.ItemStack
import gg.mineral.api.math.EulerAngle
import lombok.Getter
import lombok.RequiredArgsConstructor
import java.util.*

/**
 * The types of values that entity metadata can contain.
 */
@RequiredArgsConstructor
@Getter
enum class EntityMetadataType {
    BYTE(Byte::class.java),
    SHORT(Short::class.java),
    INT(Int::class.java),
    FLOAT(Float::class.java),
    STRING(String::class.java),
    ITEM(ItemStack::class.java),
    VECTOR(Vector::class.java),
    EULER_ANGLE(EulerAngle::class.java);

    private val dataType: Class<*>? = null

    val id: Byte
        get() = ordinal.toByte()

    companion object {
        fun byId(id: Int): EntityMetadataType {
            return entries[id]
        }
    }
}